File: 算法/代码随想录算法PDF/00调试源代码/19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第 N 个结点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummyHead = ListNode()
        dummyHead.next = head
        left, right = dummyHead, dummyHead
        while n:
            right = right.next
            n -= 1
        right = right.next # 右指针继续移动
        while right:
            right = right.next
            left = left.next
        left.next = left.next.next
        return dummyHead.next
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/19.删除链表的倒数第-n-个结点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummyHead = new ListNode(0);  // 虚拟头节点
        dummyHead->next = head;
        ListNode* right = dummyHead, *left = dummyHead;
        while (n-- && right != nullptr) {  // 可以不加判断，因为条件里有1 <= n <= sz
            right = right->next;
        }
        right = right->next;
        while (right) {
            right = right->next;
            left = left->next;
        }
        ListNode* del = left->next;
        left->next = left->next->next;
        delete del;
        return dummyHead->next;
    }
};
// @lc code=end

